PATCH-5: Add TC03

yenleagest opened the pull request.

catssaymeow committed:
PATCH-5: Add TC03

yenleagest commented:
Hi @nbmhoang, 
Could you please review this pull request? Thanks in advance.

nbmhoang commented on src/main/java/data/models/Hotel.java:
```diff
@@ -5,12 +5,20 @@
     int rating;
+    List<String> benefits;
+
+    public boolean equals(Hotel other) {
```
1. You can't override equals method without override hashcode method
2. Use EqualsAndHashCode annotation from Lombok

yenleagest replied:
1. I didn’t override equals(Object obj) because my custom method uses the signature equals(Hotel other). Since this doesn’t match Object.equals(), Java doesn’t treat it as an override:
![image](https://github.com/user-attachments/assets/f9d0a22d-121c-4980-b672-45c2b5033f0e)

2. No we cant because of these reasons:
- On the **search result page**  the destination field shows a partial address which provides the estimation of the distance from the hotel to the center of the location we search for. On the **hotel details page** it show the full address (which is expected). So a strict .equals() comparison would fail. That’s why in my custom equals() method, I use: `other.getDestination().contains(this.destination)`.
- In this PR, I added one more attribute to the hotel `List<String> benefits;`. This info is briefly listed in the search result page but again, is fully displayed in the hotel details page, as shown below. 
![Screenshot 2025-06-21 at 23 09 30](https://github.com/user-attachments/assets/601dd3f1-3ec6-492e-aae0-adb43c0dc47d)
![Screenshot 2025-06-21 at 23 09 53](https://github.com/user-attachments/assets/36f09973-15dc-4a3d-8552-f239020d5a40)
- So instead of trying to collect partial data and do a `contains` assert I leave benefits = null in the search result page and do assert the list on the hotel details page contains the input benefits.
`softAssert.assertTrue(hotelDetails.getBenefits().containsAll(data.getBenefits()));`

Does it make sense to you? And this also addresses your last comment.

yenleagest replied:
For more details, this is the Hotel object we get from the search result page:
```Hotel(name=Muong Thanh Luxury Da Nang hotel, destination=Phước Mỹ, Da Nang, price=1644681, rating=5, benefits=null)```
and this is the one on the hotel details page:
```Hotel(name=Muong Thanh Luxury Da Nang hotel, destination=No 270, Vo Nguyen Giap Street, My An ward, Ngu Hanh Son District, Da Nang city, Vietnam, Phước Mỹ, Da Nang, Vietnam, 550000, price=1644681, rating=5, benefits=[Room service [24-hour], Coffee shop, Hotel room service app, Salon, Fitness center, Spa, Swimming pool [indoor], Swimming Pool, Kids club, Tennis court, Beach, Free Wi-Fi, Pool with view, Free parking, Spa, Front desk [24-hour], Fitness center, Restaurant])```

I believe that you can see the diff and why I cant use EqualsAndHashCode

nbmhoang replied:
I see

nbmhoang commented on src/main/java/pages/agoda/HotelDetailsPage.java:
```diff
@@ -0,0 +1,88 @@
+    @Step("Get hotel's destination")
+    private String getDestination() {
+        String dest = $(destination).shouldBe(visible).getText().trim();
+        if (dest.isEmpty()) {
```
Can we use $(destination).shouldNotHave(Condition.text(""))?;

yenleagest replied:
We cant, cos the destination contains multiple spaces, used `Condition.exactText("")` instead.

nbmhoang commented on src/main/java/pages/agoda/HotelDetailsPage.java:
```diff
@@ -0,0 +1,88 @@
+
+    @Step("Get hotel's benefits")
+    private List<String> getBenefits() {
+        List<String> benefitsList = new ArrayList<>();
```
Use Stream API please!

yenleagest replied:
Done.

nbmhoang commented on src/test/java/testcases/agoda/HotelDetailsPageTest.java:
```diff
@@ -0,0 +1,93 @@
+
+        // The hotel detailed page is displayed with correct info
+        hotelDetails = hotelDetailsPage.getHotel();
+        softAssert.assertTrue(hotels.getFirst().equals(hotelDetails));
```
huh? Why dont we use assertEquals here :joy:

yenleagest replied:
Please see my reply for your first comment 👆🏼

nbmhoang reviewed (changes requested):
Thanks for the work! I’ve added a few comments — please take a look.

catssaymeow committed:
Fix review comments

yenleagest commented:
1. I also noticed that the order of items in the review details isn’t consistent across all hotels, since they’re sorted by score in descending order. That’s why I updated the validation to:
![Screenshot 2025-06-21 at 23 17 01](https://github.com/user-attachments/assets/6a98f31f-f595-4109-9110-b0724c231ac5)
`softAssert.assertTrue(reviewDetails.containsAll(data.getReviewDetails()));`

2. I just added a few methods to handle the navigation to target month before picking checkin and checkout date. Currently, Agoda displays next month by default and my codes recently broke.

catssaymeow committed:
Move Date Picker to target month before picking

nbmhoang commented on src/main/java/pages/agoda/HomePage.java:
```diff
@@ -220,4 +228,21 @@ private void closeGoogleIframe() {
         }
     }
+
+    private void moveDatePickerToMonth(LocalDate localDate) {
```
this method should be inside selectDate method and get a better name for it

yenleagest replied:
You're right. Updated!

nbmhoang commented on src/main/java/pages/agoda/HomePage.java:
```diff
@@ -220,4 +228,21 @@ private void closeGoogleIframe() {
+        }
+    }
+
+    private String getDatePickerValue() {
```
return type should be LocalDate

yenleagest replied:
The value just contains month and year, updated to YearMonth.
`<div class="DayPicker-Caption DayPicker-Caption-Wide" role="heading" aria-level="2">June 2025</div>`

nbmhoang commented on src/main/java/pages/agoda/HomePage.java:
```diff
@@ -220,4 +228,21 @@ private void closeGoogleIframe() {
+    private void moveDatePickerToMonth(LocalDate localDate) {
+        String datePickerValue = getDatePickerValue();
+        String timeRelation = DateTimeUtils.getTimeRelation(datePickerValue, localDate);
+        while (!timeRelation.equals("current")) {
```
then use isAfter/isBefore here

yenleagest replied:
Done.

nbmhoang commented on src/main/java/utils/DateTimeUtils.java:
```diff
@@ -0,0 +1,23 @@
+
+public class DateTimeUtils {
+
+    public static String getTimeRelation(String monthYear, LocalDate localDate) {
```
Just a friendly reminder: when dealing with date/time values, avoid passing raw strings to represent them. It's unclear what date format you're expecting to parse.(same for getDatePickerValue)

yenleagest replied:
🫡 Noted!

catssaymeow committed:
Fix review comments

yenleagest commented:
Hi @nbmhoang I fixed all comments regarding date time handling. Please take a look and let me know if any.

nbmhoang commented on src/main/java/pages/agoda/HotelDetailsPage.java:
```diff
@@ -0,0 +1,81 @@
+    @Step("Get hotel's rating")
+    private int getRating() {
+        String text = $(rating).shouldBe(visible).getText().split(" ")[0];
+        return (int) Math.floor(Double.parseDouble(text));
```
why don't you use Integer.parseInt directly here?
Or is there any hotel have decimal rating like 4.5 4.3?

yenleagest replied:
Yes, I saw a hotel with rating 3.5.
`<span class="ScreenReaderOnly__ScreenReaderOnlyStyled-sc-szxtre-0 jifxIN">3.5 stars out of 5</span>`

nbmhoang replied:
Then you should change data type for rating property. Otherwise, you code is rounding the rating value here

yenleagest replied:
In fact, a 3.5 rating hotel listed when filtering for rating 3. The reason I do floor round is to simlify this case. If its matter to you then I will do the update.

yenleagest replied:
Updated. Please recheck.

nbmhoang commented on src/main/java/pages/agoda/SearchResultsPage.java:
```diff
@@ -30,7 +32,10 @@ public class SearchResultsPage extends HomePage {
     private final String finalPrice = "[data-element-name='final-price']";
     private final String destination = "[data-selenium='area-city']";
+    private final String hotelName = "[data-selenium='hotel-name']";
+    private final String hotelByName = "//h3[text()='%s']";
```
This locator could be weak. Try to combine with other attribute

yenleagest replied:
Done.

nbmhoang commented on src/test/java/testcases/agoda/HotelDetailsPageTest.java:
```diff
@@ -0,0 +1,93 @@
+        softAssert.assertTrue(hotelDetails.getBenefits().containsAll(data.getBenefits()));
+
+        // Move mouse to point of the hotel to show detailed review points
+        List<String> reviewDetails = hotelDetailsPage.showReviewDetails();
```
dont declare variable here

yenleagest replied:
My bad. Updated.

catssaymeow committed:
Fix review comments

yenleagest commented:
Hi @nbmhoang please recheck. Thank you.

catssaymeow committed:
Fix review comment
